[STEP 3-1] 날씨 앱 프로젝트 (Miro, Sunny)

longlivedrgn opened the pull request:
> 안녕하세요 제이슨!(@ehgud0670) 미로와 써니 입니다 😄 저희가 Step2와 Step3의 PR을 같이 쓰기도 하고, 코드 구현 이외에도 코드 설계 측면에서 고민을 계속 하다 보니 PR이 정말 늦어지게 되었어요.😭 공부를 하던 중 보았던 아티클에 영감을 받아서 아키텍쳐까지 고려하는 코드를 작성하는데 생각보다 **정말**! 어려운 일이더라구요. 이런 방식으로 코드를 구현하는 것이 처음이라 우여곡절이 많았고 그 과정을 최대한 PR에 담아보려 노력했어요. 

현재 Step 3를 다 구현하지는 못했어요🥲 Collection View에 연결할 데이터 모델을 만드는 것까지 완성 후, Collection View를 구현하는 과정 중인데 main 브랜치에 따로 커밋은 따로 올리지 않았어요!(다른 브랜치를 보시면 저희가 고민한 과정을 보실 수 있어요..😰) Step 3의 남은 부분은 추후에 완성 후, 또 다른 PR로 올려볼게요!
아래에 저희가 Step2,3를 구현하면서 중점적으로 생각했던 요소들을 정리했어요! 이 부분을 자세히 봐주시면 감사하겠습니다🥰 만약 저희가 구현하고자 한 의도와 코드가 다르다면 그 부분들을 **꼭꼭** 알려주세요! 열심히 공부해보도록 하겠습니다.🤨

</br>

# 요구사항 (STEP2 + STEP3-1) / 핵심 구현사항
- CoreLocation을 통해 현재 좌표 가져오기
- CoreLocation을 통해 주소 가져오기
- URLSession을 통해 API 호출 (HTTP GET 메서드 명시적 호출)
- 전달받은 JSON 데이터 파싱 및 모델에 매칭
- Collection View에 사용할 데이터 모델 생성

</br>


# 고민과 해결🤔
## ViewController를 가볍게 만들기 위해 시작된 기나긴 리팩토링 여정
### 1. 문제 발생 : 거대해진 ViewController

- 기존의 ViewController의 경우, CoreLocation을 통하여 위치를 받아오기 위해 **CLLocationManagerDelegate**을 채택했어요. 
- API 통신으로 받아온 데이터를 Collection View에서 보여주기 위한 로직을 담고있어야했어요. 

```swift
// 기존 ViewController 코드
class WeatherViewController: UIViewController {

    private let networkModel = NetworkModel()
    private lazy var network = WeatherAPIManager(networkModel: networkModel)
    private let locationDelegate = LocationManagerDelegate()	    
    
    lazy var coreLocationManger: CLLocationManager = {	
        let manager = CLLocationManager()	
        manager.desiredAccuracy = kCLLocationAccuracyKilometer	
        manager.requestWhenInUseAuthorization()	
        return manager	
    }()	


    override func viewDidLoad() {	    
        super.viewDidLoad()	   

        coreLocationManger.delegate = locationDelegate	
        network.fetchWeatherInformation(
            of: .currentWeather, 
            in: Coordinate(longitude: 126.96368972, latitude: 37.53361968)
        )	
    }
}

```

- 그러던 중, [뷰 컨트롤러는 죄가 없다](https://www.sungdoo.dev/programming/do-not-blame-mvc-1) 라는 아티클을 보게되었어요. 
해당 아티클에서 쓰여진 말처럼, 저희의 코드는 네트워크 API콜을 하고 response를 받아 View에 띄우는 것 까지 모두 View Controller이 담당하고 있었어요.

</br>

### 2. 문제 해결 시도 : Data Controller의 생성
- 문제를 인식한 뒤, View Controller가 View에 데이터를 present해주는 로직만 담기로 결심했어요. 
View들을 관리하는 View Controller처럼, Data들을 관리하는 **Data Controller** 파일을 생성하기로 했어요. 
네트워크 API를 호출하고, 결과를 받아 View에 보여주기 위한 타입으로 변환하는 로직 등 데이터 관련 로직을 Data Controller에서 구현했어요. 

- Apple Documentation의 [Implementing Modern Collection Views](https://developer.apple.com/documentation/uikit/views_and_controls/collection_views/implementing_modern_collection_views) 예제 프로젝트를 참고했어요. 
저희는 이 프로젝트의 MountainsController 파일이 Data Controller 역할을 한다고 생각했어요. MountainsController는 Collection View에 데이터를 띄워주기 위해 Mountains 데이터 셋을 만들고 있어요. 

- 따라서, 저희는 ViewController와는 다른 Data Controller 파일을 생성하게 되었어요. 
Data Controller 파일에서는 CoreLocation의 알림을 받고, API 통신과 관련된 로직이 포함되어 있어요. 또한 Collection View에서 사용할 FiveDaysForecast 구조체와 CurrentWeather 구조체가 존재해요. 

</br>

### 3. 새로운 문제 발생 : 가벼운 ViewController, 거대한 Data Controller (MVVM..🤔)
- 리팩토링을 통해 가벼운 View Controller를 얻었지만, 그에 반해 매우 많은 로직을 담은 Data Controller를 만들게 되었어요...

```swift
final class WeatherController {
    
    struct CurrentWeather: Identifiable {
        let id = UUID()
        let image: UIImage?
        let address: String
        let temperatures: Temperature
    }

    struct FiveDaysForecast: Identifiable {
        let id = UUID()
        var image: UIImage?
        let date: String
        let temperature: Double
    }
    
    private var weatherAPIManager: WeatherAPIManager?
    private let locationManager = LocationManager()
    
    weak var currentWeatherDelegate: CurrentWeatherDelegate?

    var currentWeather: CurrentWeather?
    var forecaseWeather: [FiveDaysForecast] = []

    init(networkModel: NetworkModel = NetworkModel(session: URLSession.shared)) {
        weatherAPIManager = WeatherAPIManager(networkModel: networkModel)

        locationManager.locationDelegate = self
    }

    // CLLocation -> Coordinate 변환 함수
    func makeCoordinate(from location: CLLocation) -> Coordinate {
        // ...
    }
    
    // API 통신 후 decoded 한 CurrentWeatherDTO를 CurrentWeather 타입으로 변환하는 함수
    private func makeCurrentWeather(location: CLLocation) {
         // ...
    }

    // API 통신 후 decoded 한 FiveDaysForecastDTO를 FiveDaysForecast 타입으로 변환하는 함수
    private func makeForecastWeather(location: CLLocation) {
        // ... 
    }
}

    func makeWeatherData(location: CLLocation) {
        makeCurrentWeather(location: location)
        makeForecastWeather(location: location)
    }

}

extension WeatherController: LocationDelegate {
    func send(location: CLLocation) {
        makeWeatherData(location: location)
    }

}
```


- 그리고 아키텍쳐에 대한 고민이 익숙치 않아 다른 분들에게 의견을 구한 결과, 이러한 구조가 일종의 **MVVM**라는 것을 알게되었어요.
    - 사소한 궁금증인데, 보통 프로젝트 내에서 ViewModel 역할을 하는 타입의 네이밍에 "ViewModel"이라는 단어를 명시적으로 사용하나요? 😮


- 그리하여 기존의 Data Controller 파일을 View Model로 변경하고, 거대해진 View Model을 다시 가볍게 나눠보기로 했어요. 

</br>

### 4. 문제 해결 시도 : SRP를 지킨 View Model
- **SRP**란, 클래스는 하나의 기능만 가지며, 클래스가 제공하는 모든 서비스는 하나의 책임을 수행하는데 집중되어야 한다는 SOLID 원칙 중 하나에요. 
- 기존 코드는 WeatherController 안에 CurrentWeather와 FiveDaysForecast가 존재하고, DTO 데이터를 각 타입으로 변환하는 두 개의 서비스를 갖고 있어요. 이러한 WeatherController는 SRP 원칙을 위배하는 코드라고 생각했어요.


**CurrentWeatherViewModel.swift**
```swift
final class CurrentWeatherViewModel {
    
    struct CurrentWeather: Identifiable {
        let id = UUID()
        let image: UIImage?
        let address: String
        let temperatures: Temperature
    }
    
    func makeCurrentAddress(locationManager: CoreLocationManager,
                            location: CLLocation,
                            completion: @escaping (String) -> Void
    ) {
        // ...
    }
    
    func makeCurrentInformation(weatherAPIManager: WeatherAPIManager?,
                                coordinate: Coordinate,
                                location: CLLocation,
                                address: String,
                                completion: @escaping (String, CurrentWeatherDTO) -> Void
    ) {
        // ...
    }
    
    func makeCurrentImage(weatherAPIManager: WeatherAPIManager?,
                          iconString: String,
                          address: String,
                          weatherData: CurrentWeatherDTO
    ) {
             // ...
            // 추후 WeatherViewModel로 전달하는 Delegate 구현
        }
    }
}
```
- CurrentWeather 타입과 그에 관련된 서비스만을 다루는 CurrentWeatherViewModel 파일이에요.

</br>

**FiveDaysForecastWeatherViewModel.swift**
```swift
final class FiveDaysForecastWeatherViewModel {
    
    struct FiveDaysForecast: Identifiable {
            let id = UUID()
            var image: UIImage?
            let date: String
            let temperature: Double
        }
    
    func makeForecastWeather(weatherAPIManager: WeatherAPIManager?,
                                 coordinate: Coordinate,
                                 location: CLLocation,
                                 completion: @escaping (String, Day) -> Void
        ) {
             // ...
        }
        
        func makeForecastImage(weatherAPIManager: WeatherAPIManager?,
                               icon: String,
                               eachData: Day
        ) {
                // ...
                // 추후 WeatherViewModel로 전달하는 Delegate 구현

            }
        }
}
```
- FiveDaysForecast 타입과 그에 관련된 서비스만을 다루는 FiveDaysForecastWeatherViewModel 파일로 분리했어요.

</br>

**WeatherViewModel.swift**
```swift
final class WeatherViewModel {
    
    private let fiveDaysForecastWeatherViewModel = FiveDaysForecastWeatherViewModel()
    private let currentWeatherViewModel = CurrentWeatherViewModel()
    
    private let locationManager = CoreLocationManager()
    private let weatherAPIManager: WeatherAPIManager?
    
    var fiveDaysForecastWeather: [FiveDaysForecastWeatherViewModel.FiveDaysForecast] = []
    var currentWeather: CurrentWeatherViewModel.CurrentWeather?
    
    init(networkModel: NetworkModel = NetworkModel(session: URLSession.shared)) {
        weatherAPIManager = WeatherAPIManager(networkModel: networkModel)
        
        locationManager.locationDelegate = self
    }
    
    func makeCoordinate(from location: CLLocation) -> Coordinate {
        // ...
    }
    
    func makeWeatherData(locationManager: CoreLocationManager, weatherAPIManager: WeatherAPIManager?) {
        
        // currentWeather 만드는 메서드 호출
        currentWeatherViewModel.makeCurrentAddress(
            locationManager: locationManager,
            location: location
        ) { [weak self] address in
            
            self?.currentWeatherViewModel.makeCurrentInformation(
                weatherAPIManager: weatherAPIManager,
                coordinate: coordinate,
                location: location,
                address: address
            ) { [weak self] iconString, weatherData in
                
                self?.currentWeatherViewModel.makeCurrentImage(
                    weatherAPIManager: weatherAPIManager,
                    iconString: iconString,
                    address: address,
                    weatherData: weatherData
                )
            }
        }
        
        // fiveDaysForecast 만드는 메서드 호출       
        self.fiveDaysForecastWeatherViewModel.makeForecastWeather(
            weatherAPIManager: weatherAPIManager,
            coordinate: coordinate,
            location: location
        ) { [weak self] iconString, eachData in
            
            self?.fiveDaysForecastWeatherViewModel.makeForecastImage(
                weatherAPIManager: weatherAPIManager,
                icon: iconString,
                eachData: eachData
            )
        }
    }
}

extension WeatherViewModel: LocationDelegate {
    func didUpdateLocation() {
        makeWeatherData(
            locationManager: locationManager,
            weatherAPIManager: weatherAPIManager
        )
    }
}
```
- 또한 변환한 데이터 셋을 가지며 View Controller와 통신하는 WeatherViewModel 타입을 만들었어요.
- 추후 WeatherViewModel의 fiveDaysForecastWeather과 currentWeather 속성들의 값이 변경되면, ViewController가 CollectionView를 업데이트 시키는 함수 호출할 수 있도록 Delegate 구현 예정이에요.
- 이렇게 길고 길었던 리팩토링이 끝났습니다! 저희는 MVVM의 구조로(물론 처음부터 설계하고 구현한 것은 아니지만,) 코드를 구현해 보았는데 제이슨이 보시기에 MVVM 구조로 적절하게 코드를 작성한 것인지, MVVM에 대해 잘못 이해한 것은 없었는지 알려주시면 감사하겠습니다😌 

![2023-03-31_10 25 02](https://user-images.githubusercontent.com/85781941/229262090-5519607e-4eba-4051-ac63-7691f82033f8.png)

</br>


## 비동기 작업과 completionHandler 지옥(feat. async, await)
- 데이터 모델을 만들기 위해서는 기본적으로 여러 번의 API 호출을 거쳐야됐어요. Current Weather의 경우, 아래의 순서를 거쳐서 객체가 생성되요.
    1. CLGeocoder().reverseGeocodeLocation 메서드를 사용해 현재의 주소를 받아와요.
    2. 경도, 위치 정보를 통하여 CurrentWeatherDTO를 생성해요.
    3. CurrentWeatherDTO의 icon 프로퍼티 값을 통하여 icon Image를 받아오는 API 통신을 해요.
    4. 최종적으로 CurrentWeather 객체를 생성해요.
- 이러한 로직을 통하여 코드를 구현하니 아래와 같은 코드가 생성되었어요.
```swift
func makeCurrentWeather(location: CLLocation) {
        
        // 1. coordinate 주소 가져오기
        let coordinate = self.makeCoordinate(from: location)
        
        // 2. address 생성하기
        locationManager.changeGeocoder(location: location) { [weak self] place in
            
            guard let locality = place?.locality, let subLocality = place?.subLocality else { return }
            
            let address = "\(locality) \(subLocality)"
            
            // 3. currentWeather 가져오기
            self?.weatherAPIManager?.fetchWeatherInformation(of: .currentWeather, in: coordinate) { [weak self] data in
                let group = DispatchGroup()
                guard let weatherData = data as? CurrentWeatherDTO else { return }
                
                guard let icon = weatherData.weather.first?.icon else { return }
                
                group.enter()
                // 4. 이미지 가져오기
                self?.weatherAPIManager?.fetchWeatherImage(icon: icon) { [weak self] weatherImage in
                    
                    self?.currentWeather = CurrentWeather(image: weatherImage, address: address, temperatures: weatherData.temperature)
                    group.leave()
                }
                group.notify(queue: .main) {
                    self?.currentWeatherDelegate?.notifyToUpdateCurrentWeather()
                }
            }
        }
    }
```
- 연속적인 completionHandler로 인해 코드의 가독성이 좋지 못하다고 생각했어요.
- 이러한 장풍 코드를 개선하기 위한 방법을 고민해봤어요.
    - DispatchGroup의 활용해 return 값으로 데이터를 전달한다.
    - async, await를 활용한다.
    
    먼저, DispatchGroup의 경우 여러 개의 DispatchGroup를 활용하다보니, enter(), leave()가 꼬이게 되었고 비동기 작업이 정상적으로 진행되지 않음을 알 수 있었어요.

    두 번째, `async`, `await`의 경우 공부해보지 못한 개념이라 코드에 적용하지 못했어요. 공식문서로 잠깐 보았던 async, await는 completionHandler보다 훨씬 좋은 가독성을 보이는 것 같아 공부 이후 리팩토링 할 예정이에요.😎
    
    따라서, 저희는 기존의 completionHandler를 여러 개의 함수로 나누기로 결정했어요.

- 그리하여 완성된 코드는 아래와 같아요.
```swift
// WeatherViewModel
func makeWeatherData(locationManager: CoreLocationManager, weatherAPIManager: WeatherAPIManager?) {
        
        guard let location = locationManager.locationManager?.location else { return }
        let coordinate = self.makeCoordinate(from: location)
        
        currentWeatherViewModel.makeCurrentAddress(
            locationManager: locationManager,
            location: location
        ) { [weak self] address in
            
            self?.currentWeatherViewModel.makeCurrentInformation(
                weatherAPIManager: weatherAPIManager,
                coordinate: coordinate,
                location: location,
                address: address
            ) { [weak self] iconString, weatherData in
                
                self?.currentWeatherViewModel.makeCurrentImage(
                    weatherAPIManager: weatherAPIManager,
                    iconString: iconString,
                    address: address,
                    weatherData: weatherData
                )
            }
        }
    
// CurrentWeatherViewModel
func makeCurrentAddress(locationManager: CoreLocationManager,
                        
                            location: CLLocation,
                            completion: @escaping (String) -> Void
    ) {
        locationManager.changeGeocoder(location: location) { place in
            
            guard let locality = place?.locality, let subLocality = place?.subLocality else { return }
            let address = "\(locality) \(subLocality)"
            completion(address)
        }
    }
    
    func makeCurrentInformation(weatherAPIManager: WeatherAPIManager?,
                                coordinate: Coordinate,
                                location: CLLocation,
                                address: String,
                                completion: @escaping (String, CurrentWeatherDTO) -> Void
    ) {
        weatherAPIManager?.fetchWeatherInformation(of: .currentWeather, in: coordinate) { data in
            
            guard let weatherData = data as? CurrentWeatherDTO else { return }
            guard let icon = weatherData.weather.first?.icon else { return }
            
            completion(icon, weatherData)
        }
    }
    
    func makeCurrentImage(weatherAPIManager: WeatherAPIManager?,
                          iconString: String,
                          address: String,
                          weatherData: CurrentWeatherDTO
    ) {
        weatherAPIManager?.fetchWeatherImage(icon: iconString) { weatherImage in
            
            let currentWeatherData = CurrentWeather(image: weatherImage, address: address, temperatures: weatherData.temperature)
        }
    }
```

SunnnySong committed:
refactor: ViewController에서 networkModel 생성 이후 주입

longlivedrgn committed:
feat: CoreLocation을 통하여 현재 주소 받아오고 Model에 매핑하기

SunnnySong committed:
feat: CoreLocation을 통해 가져온 좌표를 주소값으로 변환하기

longlivedrgn committed:
chore: HTTP GET 메서드 명시적 호출

SunnnySong committed:
refactor: .notDetermined: 수정

SunnnySong committed:
Merge branch 'main' of https://github.com/longlivedrgn/ios-weather-forecast

SunnnySong committed:
refactor: CLLocationManagerDelegate를 ViewController에서 분리하기

SunnnySong committed:
chore: Views 폴더 생성 및 Delegate 하위 폴더 생성

longlivedrgn committed:
chore: 기존 데이터 모델 속 필요 없는 프로퍼티 제거

SunnnySong committed:
feat: WeatherController 생성

longlivedrgn committed:
refactor: task 함수 및 fetchWeatherInformation 함수 역할 조정

SunnnySong committed:
chore: DTO 관련 타입 Codable -> Decodable 변경

longlivedrgn committed:
feat: ImageURL을 생성하는 static method 생성

SunnnySong committed:
chore: WeatherViewController 불필요한 코드 삭제 및 Network 폴더링

longlivedrgn committed:
feat: FiveDaysForecastWeatherViewModel, CurrentWeatherViewModel 파일 추가

SunnnySong committed:
feat: WeatherViewModel 생성

longlivedrgn committed:
feat: CoreLocationManager 및 locationDelegate 구현 및 생성

SunnnySong committed:
feat: WeatherAPIManager에 WeatherImage 로딩 메서드 구현, CurrentWeatherViewModel 에 networking 코드 구현

longlivedrgn committed:
feat: FiveDaysForecastWeatherViewModel 속 makeForecastWeather, makeForecastImage 메소드 생성

SunnnySong committed:
feat: WeatherViewModel 생성 및 LocationDelegate 연결

longlivedrgn committed:
feat: viewcontroller에 weatherViewModel 생성

SunnnySong committed:
feat: WeatherViewController에 WeatherViewModel 주입

ehgud0670 commented on WeatherForecast/WeatherForecast/ViewModels/WeatherViewModel.swift:
```diff
@@ -0,0 +1,84 @@
+        return Coordinate(longitude: longitude, latitude: latitude)
+    }
+    
+    func makeWeatherData(locationManager: CoreLocationManager, weatherAPIManager: WeatherAPIManager?) {
```
make 라고 이름 짓는 메소드는 보통 무언갈 만들어서 return 해주는 메소드입니다. 따라서 return type 이 void 인데 메소드 이름이 make인 메소드들은 다른 동료에게 오해를 불러 일으킬수 있으므로 다른 네이밍이 좋아 보여요~

longlivedrgn replied:
제이슨의 피드백을 참고해서 아래와 같이 구조와 각 타입의 이름을 변경해보았어요!🧐

1️⃣ `URLSessionDataTask`를 생성하는 `NetworkModel`을 `NetworkSession`으로 이름을 변경했어요
2️⃣ `NetworkSession`이 담당하는 decode 메소드를 `JSONNetworkDeserializer`라는 타입을 생성하여 담당하게 해주었어요.
3️⃣ `WeatherAPIManager`를 `WeatherNetworkDispatcher`로 변경하여 **requestWeatherInformation** 메소드를 통하여 DataTask를 실행하도록 하였습니다!
4️⃣ `FiveDaysForecastWeatherViewModel`, `CurrentWeatherViewModel`는 앞선 `WeatherNetworkDispatcher`를 통하여 Information을 fetch해오는 **fetchInformation**메소드가 존재해요.
5️⃣ `WeatherViewModel`는 각 ViewModel의 Network 메소드를 실행시키는 **execute** 메소드가 존재해요. 
- 추후에 testable하게 Network 코드를 리팩토링해볼 예정입니다 ☺️

ehgud0670 commented on WeatherForecast/WeatherForecast/Network/Model/NetworkModel.swift:
```diff
@@ -7,14 +7,16 @@
-typealias NetworkResult = Result<Decodable, NetworkError>
+typealias NetworkResult = Result<Data, NetworkError>
 
 final class NetworkModel {
```
NetworkModel이라는 이름보다는 NetworkManager 라는게 더 어울릴 것 같아요. 
Model이라고 지으니 프로퍼티만 있고 메소드는 없어보이는 객체같네요!

ehgud0670 replied:
음 상위객체에 WeatherAPIManager라는게 있어서 NetworkManager라고 이름지으면 중복이 되네요!. 그럼 이 객체 이름은 NetworkDispatcher 라고 이름짓는것도 괜찮을 것 같아요.
Network Request를 보내고 받아서 디코딩하는, 네트워크에서 가장 낮은 수준의 하위 객체인거죠.
참고한 링크: https://danielbernal.co/2021/01/11/writing-a-networking-library-with-combine-codable-and-swift-5/, 
https://medium.com/ios-os-x-development/minimal-networking-layer-from-scratch-in-swift-4-a151af786dc5

longlivedrgn replied:
https://github.com/tasty-code/ios-weather-forecast/pull/12#discussion_r1155720192
🙄 위의 코멘트에 해당 코멘트를 작성해보았어요!

ehgud0670 commented on WeatherForecast/WeatherForecast/Model/CoreLocationManager.swift:
```diff
@@ -0,0 +1,65 @@
+
+final class CoreLocationManager: NSObject {
+    
+    weak var locationDelegate: LocationDelegate?
```
CoreLocationManager 에 대한 델리게이트이므로 이런식으로 작명하시는 것이 코드 이해하는데 더 좋을 것 같아요.
Delegate를 쓰는곳에선 이런식으로 보통 네이밍해서요.

```suggestion
    weak var delegate: CoreLocationManagerDelegate?
```

SunnnySong replied:
넵 수정했습니다!😃

ehgud0670 commented on WeatherForecast/WeatherForecast/Model/CoreLocationManager.swift:
```diff
@@ -0,0 +1,65 @@
+final class CoreLocationManager: NSObject {
+    
+    weak var locationDelegate: LocationDelegate?
+    var locationManager: CLLocationManager?
```
```suggestion
    private var locationManager: CLLocationManager?
```
private으로 감추는것이 좋을 것 같아요 그리고 항상 init 될때마다 생성되어야 하는 프로퍼티라면 non-optional 인게 더 좋을 것 같아요.

longlivedrgn replied:
locationManager의 타입을 non-optional 타입으로 바꾸고, private으로 감추는 대신에 현재 위치를 return하는 메소드인 fetchCurrentLocaiton()를 추가하였습니다.

ehgud0670 commented on WeatherForecast/WeatherForecast/Model/LocationDelegate.swift:
```diff
@@ -0,0 +1,14 @@
+import Foundation
+import CoreLocation
+
+protocol LocationDelegate: AnyObject {
+    func didUpdateLocation()
+}
```
```suggestion
protocol CoreLocationManagerDelegate: AnyObject {
    func coreLocationManager(_ manager: CoreLocationManager, didUpdateLocations locations: [CLLocation])
}
```
보통 델리게이트 메소드의 경우는 보통 위처럼 델리게이트에 대한 source를 첫번째 파라미터로 만들고 기타 넘길 필요가 있는 값을 추가로 파라미터로 만듭니다. 참고하시고 왜 보통 그런지 이유도 생각해보시면 좋을 것 같아요!

SunnnySong replied:
제이슨의 리뷰를 받고 Swift Style Guide에서 찾아본 delegate naming 관련 내용을 찾아 보았어요. 찾아본 결과 delegate methods에서는 delegate's source object를 첫번째 파라미터로 사용한다 하더라구요. 수정 완료했습니다!

ehgud0670 reviewed (commented):
고생하셨습니다 :)
주로 스타일 가이드에 대해서만 코멘트 남겨두었습니다.
* WeatherViewModel을 결국엔 쓰지 않은게 아쉽네요. 결국에 WeatherViewModel 코드가 쓰일 때 해당 뷰모델에 대한 코멘트도 할수 있을 것 같아요 🙂
* 그리고 리뷰어가 빌드를 해서 디버깅하거나 UI를 테스트 할 수 있어야 해서 APIKeys.swift들도 지금처럼 리뷰받는 환경에서는 .gitignore에 포함을 안시키는 게 훨씬 좋을 듯 합니다 :)
추가로 .gitignore로 보안이 필요한 파일을 감추는 방법 말고도 Info.plist에 APIKey 값을 두는 방법등([관련해서 참고하면 좋을 내용](https://medium.com/swift-india/secure-secrets-in-ios-app-9f66085800b4))이 있으니 추가로 공부하시면 좋을 것 같아요 그럼 고생 많으셨습니다!

ehgud0670 commented on WeatherForecast/WeatherForecast/ViewModels/FiveDaysForecastWeatherViewModel.swift:
```diff
@@ -0,0 +1,42 @@
+            let temperature: Double
+        }
+    
+    func makeForecastWeather(weatherAPIManager: WeatherAPIManager?,
```
결국 apiManager를 통해 fetch 하고 있으니 make라는 메소드 명 말고 get. request, execute 와 같은 서버 통신을 한다는 걸 알수 있는 네이밍이 어떨까요?

longlivedrgn replied:
https://github.com/tasty-code/ios-weather-forecast/pull/12#discussion_r1155720192
🙄 이것 또한 위의 코멘트에 해당 코멘트를 작성해보았어요!

ehgud0670 commented on WeatherForecast/WeatherForecast/ViewModels/WeatherViewModel.swift:
```diff
@@ -0,0 +1,84 @@
+    private let currentWeatherViewModel = CurrentWeatherViewModel()
+    
+    private let locationManager = CoreLocationManager()
+    private let weatherAPIManager: WeatherAPIManager?
```
이 객체도 init 때 반드시 생성되는 객체인데 옵셔널 타입으로 둘 필요가 있을까요? 🤔

SunnnySong replied:
습관적으로 옵셔널을 사용해 속성을 정의했던 것 같아요. 앞으로 주의하겠습니다!

ehgud0670 commented on WeatherForecast/WeatherForecast/ViewModels/WeatherViewModel.swift:
```diff
@@ -0,0 +1,84 @@
+        locationManager.locationDelegate = self
+    }
+    
+    func makeCoordinate(from location: CLLocation) -> Coordinate {
```
coordinate는 결국에 이 객체 내부에서만 사용할 것 같아요 따라서 해당 메소드는 private 으로 감추는게 나을 것 같아요.
객체 구현하실때 한 모듈안에 다른 파일에 공개되어야 하는 메소드는 internal로, 감출 필요가 있는 메소드는 private 으로 두는 것이 좋고 설계 때부터 접근 제한을 어떻게 둘지 고민하시면 좋습니다 👍

SunnnySong replied:
설계를 최대한 완벽히 하고 나서 구현하는 순서로 프로젝트를 진행하려 노력하고 있지만 생각보다 더 어려워서 노력하고 있습니다.😭 앞으로 접근 제한에 대해 설계 단에서 한번 더 생각하고 코드를 구현하도록 하겠습니다!

ehgud0670 reviewed (commented):
고생하셨습니다!

SunnnySong committed:
refactor: CoreLocationManager 리뷰 수정

- locationManager 속성에 private 붙이기
- CLLocationManager.location 얻는 함수 생성하기

SunnnySong committed:
Merge branch 'main' of https://github.com/longlivedrgn/ios-weather-forecast

longlivedrgn committed:
refactor: APIKeys.swift .gitignore에서 제거

longlivedrgn committed:
refactor: delegate 변수명 변경

- Convention에 맞게 delegate 변수명 변경

SunnnySong committed:
refactor: CoreLocationManagerDelegate 네이밍 수정

- LocationDelegate -> CoreLocationManagerDelegate로 네이밍 수정
- delegate 함수 명을 coreLocationManager로 수정
- coreLocationManager의 파라미터로 CoreLocationManager을 추가해주면서 CoreLocationManager 인스턴스를 특정지어서 전달할 수 있도록 수정

longlivedrgn committed:
refactor: NetworkModel, WeatherAPIManger 이름 변경 및 JSONDeserializer 생성

- NetworkModel을 NetworkSession으로 변경
- WeatherAPIManger를 NetworkDispatcher로 변경
- Decode를 해주는 JSONDeserializer를 생성

longlivedrgn committed:
refactor: NetworkDispatcher을 WeatherNetworkDispatcher로 변경

SunnnySong committed:
refactor: WeatherViewModel method 이름 변경

- WeatherNetworkDispatcher 옵셔널 제거
- makeCoordinate private 추가
- makeWeatherData -> execute로 네이밍 변경

ehgud0670 reviewed (approved):
고생하셨습니다~ 👏
